fuel-p2p/src/discovery.rs

```diff
@@ -0,0 +1,498 @@
+                    });
+                }
+                NetworkBehaviourAction::NotifyHandler {
+                    peer_id,
+                    handler,
+                    event,
+                } => {
+                    return Poll::Ready(NetworkBehaviourAction::NotifyHandler {
+                        peer_id,
+                        handler,
+                        event,
+                    })
+                }
+                NetworkBehaviourAction::ReportObservedAddr { address, score } => {
+                    return Poll::Ready(NetworkBehaviourAction::ReportObservedAddr {
+                        address,
+                        score,
+                    })
+                }
+            }
+        }
+
+        while let Poll::Ready(mdns_event) = self.mdns.poll(cx, params) {
+            match mdns_event {
+                NetworkBehaviourAction::GenerateEvent(mdns_event) => match mdns_event {
+                    MdnsEvent::Discovered(list) => {
+                        // inform kademlia of newly discovered local peers
+                        // only if there aren't enough peers already connected
+                        if self.connected_peers_count < self.max_peers_connected {
+                            for (peer_id, multiaddr) in list {
+                                self.kademlia.add_address(&peer_id, multiaddr);
+                            }
+                        }
+                    }
+                    _ => {}
+                },
+                NetworkBehaviourAction::ReportObservedAddr { address, score } => {
+                    return Poll::Ready(NetworkBehaviourAction::ReportObservedAddr {
+                        address,
+                        score,
+                    })
+                }
+                NetworkBehaviourAction::CloseConnection {
+                    peer_id,
+                    connection,
+                } => {
+                    return Poll::Ready(NetworkBehaviourAction::CloseConnection {
+                        peer_id,
+                        connection,
+                    })
+                }
+                _ => {}
+            }
+        }
+
+        if let Some(next_event) = self.events.pop_front() {
+            return Poll::Ready(NetworkBehaviourAction::GenerateEvent(next_event));
+        }
+
+        Poll::Pending
+    }
+
+    /// return list of known addresses for a given peer
+    fn addresses_of_peer(&mut self, peer_id: &PeerId) -> Vec<Multiaddr> {
+        let mut list = self
+            .predefined_nodes
+            .iter()
+            .filter_map(|(current_peer_id, multiaddr)| {
+                if current_peer_id == peer_id {
+                    Some(multiaddr.clone())
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        {
+            let mut list_to_filter = Vec::new();
+
+            list_to_filter.extend(self.kademlia.addresses_of_peer(peer_id));
```

Comment:
can't we just extend `list` here?

Reply:
no, I need to filter that list first - thanks for reminding me tho 👍